### PR TITLE
Generic FlowContainer.

### PR DESCRIPTION
**Pull request by peppy**
open to feedback.

can potentially make an `internal` interface to keep some of the stuff non-public if necessary.

### CODE REVIEW

**Comment by peppy** [comment]
also for simplicity, we may want to convert all containers to `Container<Drawable>` and remove the `Container` class.

**Comment by SirCmpwn** [comment]
I don't understand the value-add of having Container be generic. You didn't change anything that uses the Drawable to use T instead, like Add or Children or anything at all, really.

**Comment by peppy** [comment]
`FlowContainer<BeatmapPanel>`

`FlowContainer<OptionsItem>`

`FlowContainer<Button>`

think of all the cases this reduces the need for `.Children.Cast()` or a second list storing stuff for actual logic use.

**Comment by SirCmpwn** [comment]
But it doesn't? Children is still an IEnumerable<Drawable>, not an IEnumerable<T>

File: osu.Framework/Graphics/Containers/IContainer.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) 2007-2016 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu-framework/master/LICENCE
+
+using System;
+using System.Collections.Generic;
+using OpenTK;
+
+namespace osu.Framework.Graphics.Containers
+{
+    public interface IContainer : IDrawable
```
Comment (Tom94) [inline osu.Framework/Graphics/Containers/IContainer.cs]: I'd rather not split up IContainer and have the entire interface in IContainer<T>. I don't see any good reason why someone would want to inherit from just IContainer.

File: osu.Framework/Graphics/Containers/Container.cs [hunk 2]
```diff
@@ -5,22 +5,22 @@
...
 using osu.Framework.Graphics.Primitives;
 using osu.Framework.Graphics.OpenGL;
 using OpenTK.Graphics;
 using osu.Framework.Graphics.Shaders;
-using osu.Framework.Graphics.Sprites;
-using System.Threading.Tasks;
 
 namespace osu.Framework.Graphics.Containers
 {
+    public class Container : Container<Drawable>
```
Comment (Tom94) [inline osu.Framework/Graphics/Containers/Container.cs]: I feel like we should in fact not have this, given that we can no longer cast e.g. `FlowContainer` to `Container` in such a scenario.

Comment (peppy) [inline osu.Framework/Graphics/Containers/Container.cs]: yeah, wanted to ask your opinion on it. i was thinking the same thing.

File: osu.Framework/Graphics/Containers/IContainer.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+using System;
+using System.Collections.Generic;
+using OpenTK;
+
+namespace osu.Framework.Graphics.Containers
+{
+    public interface IContainer
+    {
+        Vector2 ChildSize { get; }
```
Comment (Tom94) [inline osu.Framework/Graphics/Containers/IContainer.cs]: My old comment got overwritten by your changes: I still would prefer if all `IContainer` members would just be moved into `IContainer<T>`. I don't think there is merit to splitting it up into 2 interfaces. Rather it allows people to do weird things like inherit `IContainer` but not `IContainer<T>`.

**Pull request merged**